tests/pass-dep/libc/libc-socketpair.rs

```diff
@@ -0,0 +1,121 @@
+    assert_eq!(res, 5);
+    let mut buf3: [u8; 5] = [0; 5];
+    res = unsafe {
+        libc::read(fds[0], buf3.as_mut_ptr().cast(), buf3.len() as libc::size_t).try_into().unwrap()
+    };
+    assert_eq!(res, 5);
+    assert_eq!(buf3, "12345".as_bytes());
+
+    // Read size > data available in buffer.
+    let data = "123".as_bytes().as_ptr();
+    res = unsafe { libc::write(fds[1], data as *const libc::c_void, 3).try_into().unwrap() };
+    assert_eq!(res, 3);
+    let mut buf4: [u8; 5] = [0; 5];
+    res = unsafe {
+        libc::read(fds[0], buf4.as_mut_ptr().cast(), buf4.len() as libc::size_t).try_into().unwrap()
+    };
+    assert_eq!(res, 3);
+    assert_eq!(&buf4[0..3], "123".as_bytes());
+}
+
+fn test_socketpair_threaded() {
+    use std::thread;
+    let mut fds = [-1, -1];
+    let mut res =
+        unsafe { libc::socketpair(libc::AF_UNIX, libc::SOCK_STREAM, 0, fds.as_mut_ptr()) };
+    assert_eq!(res, 0);
+
+    let data = "abcde".as_bytes().as_ptr();
+    res = unsafe { libc::write(fds[0], data as *const libc::c_void, 5).try_into().unwrap() };
+    assert_eq!(res, 5);
+    let thread1 = thread::spawn(move || {
+        let mut buf: [u8; 5] = [0; 5];
+        let res: i64 = unsafe {
+            libc::read(fds[1], buf.as_mut_ptr().cast(), buf.len() as libc::size_t)
+                .try_into()
+                .unwrap()
+        };
+        assert_eq!(res, 5);
+        assert_eq!(buf, "abcde".as_bytes());
+    });
+    let _ = thread1.join();
+
+    // Read and write from different direction
+    let thread2 = thread::spawn(move || {
+        let data = "12345".as_bytes().as_ptr();
+        let res: i64 =
+            unsafe { libc::write(fds[0], data as *const libc::c_void, 5).try_into().unwrap() };
+        assert_eq!(res, 5);
+    });
+    let _ = thread2.join();
+    let mut buf: [u8; 5] = [0; 5];
+    res = unsafe {
+        libc::read(fds[1], buf.as_mut_ptr().cast(), buf.len() as libc::size_t).try_into().unwrap()
+    };
+    assert_eq!(res, 5);
+    assert_eq!(buf, "12345".as_bytes());
+}
+fn test_write_race() {
+    use std::thread;
+    let mut fds = [-1, -1];
+    let _res = unsafe { libc::socketpair(libc::AF_UNIX, libc::SOCK_STREAM, 0, fds.as_mut_ptr()) };
+    let thread1 = thread::spawn(move || {
+        let arr0: [u8; 100000] = [0; 100000];
+        let _ = unsafe { libc::write(fds[0], arr0.as_ptr() as *const c_void, 100000) };
+    });
+    let thread2 = thread::spawn(move || {
+        let arr1: [u8; 100000] = [1; 100000];
+        let _ = unsafe { libc::write(fds[0], arr1.as_ptr() as *const c_void, 100000) };
+    });
+    let _ = thread1.join();
+    _ = thread2.join();
+    let mut buf: [u8; 200000] = [2; 200000];
+    _ = unsafe { libc::read(fds[1], buf.as_mut_ptr().cast(), buf.len() as libc::size_t) };
+    for byte in buf.iter().take(100000) {
+        if *byte != 0 {
+            // The first 100000 bytes should all be 0.
+            println!("race happened");
+        }
+    }
+}
```

Comment:
Currently there is no synchronization on the socketpair buffers yet. I suspect a data race can happen, but it is not detected in this test. @RalfJung please let me know your thoughts on this. :)

Reply:
When a race happened, a data pattern of 00100 might appear in the buffer because the write of thread2 started before the write of thread1 finished.

Reply:
Yea, our write/read operations are atomic if they succeed fully. So if the returned value from `write` is `100000`, then all of the bytes were written, and can't have been interleaved with the other thread's written data.

The only thing that could happen is that the first 100k bytes are `1`, because it's not guaranteed which thread writes first. But there will be no interleaving.

What is the expected behaviour on a real system? Is this going to race? If so, we'll need to teach our data race detector about file descriptors.

Reply:
I think `write` calls are atomic in real systems as well.

However, for this test, can't the second thread go first so we get the 1s before the 0s?

Reply:
> But there will be no interleaving.

Yes, this is good enough. 

> I think write calls are atomic in real systems as well.

I haven't found the specification for ``write`` on ``socketpair``, but at least for ``pipes`` or ``FIFO``, ``write`` is only guranteed to be atomic when the ``bytes written`` <= ``buffer size``

> can't the second thread go first so we get the 1s before the 0s?

Yes, it is still correct even if second thread goes first. It is fine as long as the ``write`` of both threads are not interleaved (not having the ``00100`` pattern)

Reply:
The test checks that the first 10k bytes are 0, doesn't it? So if the 2nd thread goes first it will fail? I must be missing something.

Reply:
yea, the test will fail under some seeds and preemption rates

Reply:
Second thread goes first is a correct behaviour to have in a real system, but I didn't consider it in this test. Don't worry, I wrote this test is just to discuss if interleaving write is possible, I didn't plan to keep this test.